File: src/conduit/engine/transform.py
"""Transform data using DuckDB SQL engine."""

from __future__ import annotations

import logging

import duckdb
import pyarrow as pa

logger = logging.getLogger(__name__)


def run_transform(sql: str, sources: dict[str, pa.Table]) -> pa.Table:
    """Register source tables in DuckDB and execute the transform SQL.

    Returns the result as an Arrow table.
    """
    conn = duckdb.connect(":memory:")

    for name, table in sources.items():
        conn.register(name, table)
        logger.debug("Registered source '%s' (%d rows, %d cols)", name, len(table), len(table.schema))

    logger.info("Running transform SQL (%d characters)", len(sql.strip()))
    result = conn.execute(sql).to_arrow_table()
    logger.info("Transform complete: %d rows, %d columns", len(result), len(result.schema))

    conn.close()
    return result

File: src/ts/engine/transform.ts
/**
 * Transform data using DuckDB SQL engine.
 *
 * Uses the normalize module to register source DataTables as DuckDB
 * tables, then executes the user's SQL and returns the result.
 */

import type { DataTable } from "../models.ts";
import {
  createSession,
  closeSession,
  registerSources,
  querySession,
} from "../normalize/mod.ts";

/**
 * Register source tables in DuckDB and execute the transform SQL.
 * Returns the result as a DataTable.
 */
export async function runTransform(
  sql: string,
  sources: Map<string, DataTable>,
): Promise<DataTable> {
  const session = await createSession();

  try {
    // Normalize: convert all source DataTables → DuckDB tables
    await registerSources(session, sources);

    // Transform: run user SQL across all registered tables
    console.log(`Running transform SQL (${sql.trim().length} characters)`);
    const result = await querySession(session, sql);

    console.log(
      `Transform complete: ${result.rows.length} rows, ${result.columns.length} columns`,
    );
    return result;
  } finally {
    closeSession(session);
  }
}
